blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md

```diff
@@ -0,0 +1,317 @@
+s\\), and inputs \\(x_1 \in A(s_1),...,x_n \in A(s_n)\\) we should get the
+same result if we evaluate \\(A(t)(x_1,\ldots,x_n)\\) and then apply
+\\(\alpha_s\\) as if we first translate \\(x_1,\ldots,x_n\\) to
+\\(B(s_1),\ldots,B(s_n)\\) and then apply \\(B(t)\\). If you unwind the
+definitions, this is exactly what naturality says.
+
+Then we have a category of \\(T\\)-algebras and we can ask if there are
+initial or final algebra.
+It turns out that both of them *always* exist.
+
+The initial algebra is most famous here, we define for each sort
+\\(In(T)(s) = \cdot \vdash s\\), the closed terms of that sort modulo the
+equivalence of the theory, and \\(In(T)(s_1,\ldots,s_n) =
+In(T)(s_1)\times\ldots,In(T)(s_n)\\). Then the terms are just
+interpreted as the functions you get by plugging closed inputs into
+them. Then if we look at what what a morphism of \\(T\\)-algebras from
+\\(In(T) \to A\\) is, we see that we don't have any choice, the only one
+is the one that maps \\(\cdot \vdash t : s\\) to \\(A(t)\\) and this makes all
+the right diagrams to commute.
+This is pretty similar to our definition of "initial algebra" before,
+except that `count` was defined as a function, not an ADT, but that
+was just an easy way to satisfy the computational equations for
+`count`.
+
+However, an egregious flaw presents itself when we look at what the
+*final* algebra is. It's completely trivial! We can define \\(Fin(T)\\) to
+take every sort to a one element set \\(Fin(T)(s) = \{*\}\\) and every
+term to the trivial function \\(\{*\}^n \to \{*\}\\).
+What the hell?
+This interprets numbers and multisets as trivial one-element sets.
+To rule this one out, we need to add some conditions to our algebras.
+
+# Formalization: Algebras of a Theory Extension
+
+To rule out these boring algebras, and get a nice final algebra, we
+have to recognize that the sorts `num` and `multiset` in our theory
+are not really on equal footing.
+While we are not sure how multisets should be defined, we know
+*exactly* what numbers are!
+
+To formalize this we'll call the full theory \\(T_1\\) and the theory with
+just numbers \\(T_0\\). Then there should be a map from \\(T_0\\) to \\(T_1\\)
+that is the inclusion of theories.  We'll formalize this as a
+*morphism of theories*. A morphism of theories is a *strict*
+product-preserving functor from one theory to another.
+The strictness ensures that we don't mix up our sorts and our
+contexts, a morphim of theories has to map sorts to sorts, whereas a
+non-strict functor could map a sort to a context with two sorts it's
+equivalent to.
+What this really amounts to is a translation of one theory into
+another. It maps sorts to sorts and terms to terms of the appropriate
+sorts in a compositional way.
+However, we don't want to consider *all* such morphisms, only the ones
+that are "conservative extensions", i.e., there are no new closed
+terms at old types and closed terms that were different before remain
```

Comment:
suggestion: "(1) there are no new closed terms at old types and (2) closed terms ...."